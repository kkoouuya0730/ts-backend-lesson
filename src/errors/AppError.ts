export class AppError extends Error {
  constructor(public statusCode: number, public message: string) {
    super(message);
    this.statusCode = statusCode;
    this.message = message;

    Object.setPrototypeOf(this, new.target.prototype);
  }
}
