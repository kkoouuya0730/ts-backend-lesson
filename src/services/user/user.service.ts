import prisma from "../../models/prisma";
import { UserInput } from "../../validation";

export const getUsers = async () => {
  const users = await prisma.user.findMany();
  return users.map(({ passwordHash, ...publicUser }) => publicUser);
};

export const getUser = async (id: number) => {
  const user = await prisma.user.findUnique({
    where: {
      id,
    },
  });
  if (!user) return null;
  const { passwordHash, ...publicUser } = user;
  return publicUser;
};

export const createUser = async (data: UserInput) => {
  return await prisma.user.create({
    data,
  });
};

export const updateUser = async (id: number, name: string) => {
  const updatedUser = await prisma.user.update({
    where: {
      id,
    },
    data: {
      name,
    },
  });
  const { passwordHash, ...publicUser } = updatedUser;
  return publicUser;
};

export const deleteUser = async (id: number) => {
  await prisma.user.delete({
    where: {
      id,
    },
  });
};
