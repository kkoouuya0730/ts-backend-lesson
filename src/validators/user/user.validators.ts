import { ZodSafeParseResult } from "zod";
import { userIdSchema, UserInput, userInputSchema } from "../../validation";

/**
 * 例外を投げる関数を限定するという意図のもと、
 * parse系関数では例外を投げず、呼び出しもとで例外処理を行う
 */

export const parseUserId = (
  id: string | undefined
): ZodSafeParseResult<{
  id: number;
}> => {
  const parsedUserIdResult = userIdSchema.safeParse({ id });
  return parsedUserIdResult;
};

export const parseUserInput = (
  input: UserInput | undefined
): ZodSafeParseResult<{
  name: string;
}> => {
  const parsedInput = userInputSchema.safeParse(input);
  return parsedInput;
};

export const parseCreateUserInput = (
  input: UserInput | undefined
): ZodSafeParseResult<{
  name: string;
}> => {
  const parsedInput = userInputSchema.safeParse(input);
  return parsedInput;
};
