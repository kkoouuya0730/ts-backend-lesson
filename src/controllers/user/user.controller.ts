import * as userService from "../../services/user/user.service";
import { Request, Response, NextFunction } from "express";
import { AppError } from "../../errors/AppError";
import { INVALID_USER_ID, INVALID_USER_INPUT, USER_NOT_FOUND } from "../../constants";
import { userIdSchema, userInputSchema } from "../../validation";

export const getUsersHandler = async (_req: Request, res: Response, next: NextFunction) => {
  try {
    const users = await userService.getUsers();
    res.json(users);
  } catch (error) {
    next(error);
  }
};

export const getUserHandler = async (req: Request, res: Response, next: NextFunction) => {
  const parsedUserIdResult = userIdSchema.safeParse({ id: req.params.id });
  console.log(parsedUserIdResult);
  if (!parsedUserIdResult.success) {
    throw new AppError(400, INVALID_USER_ID, parsedUserIdResult.error.issues);
  }

  const userId = parsedUserIdResult.data.id;

  try {
    const user = await userService.getUser(userId);
    if (!user) {
      throw new AppError(404, USER_NOT_FOUND);
    }
    res.json(user);
  } catch (error) {
    next(error);
  }
};

export const updateUserHandler = async (req: Request, res: Response, next: NextFunction) => {
  const parsedUserIdResult = userIdSchema.safeParse({ id: req.params.id });

  if (!parsedUserIdResult.success) {
    throw new AppError(400, INVALID_USER_ID, parsedUserIdResult.error.issues);
  }

  const userId = parsedUserIdResult.data.id;

  const parsedInput = userInputSchema.safeParse(req.body);
  if (!parsedInput.success) {
    throw new AppError(400, INVALID_USER_INPUT, parsedInput.error.issues);
  }

  try {
    const user = await userService.updateUser(userId, parsedInput.data.name);
    res.json(user);
  } catch (error) {
    next(error);
  }
};

export const deleteUserHandler = async (req: Request, res: Response, next: NextFunction) => {
  const parsedUserIdResult = userIdSchema.safeParse({ id: req.params.id });
  if (!parsedUserIdResult.success) {
    throw new AppError(400, INVALID_USER_ID, parsedUserIdResult.error.issues);
  }

  const userId = parsedUserIdResult.data.id;

  try {
    await userService.deleteUser(userId);
    res.json({ message: "User deleted successfully" });
  } catch (error) {
    next(error);
  }
};
