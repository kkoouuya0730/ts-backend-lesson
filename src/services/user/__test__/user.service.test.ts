import prisma from "../../../models/prisma";
import * as userService from "../../user/user.service";
import { AppError } from "../../../errors/AppError";
import { mockUser, mockUsers, mockUsersWithNoPassword } from "../../../mocks/user.mock";

describe("User Service", () => {
  afterEach(() => {
    jest.restoreAllMocks();
  });

  describe("正常系", () => {
    test("getUsers はユーザー一覧を返す", async () => {
      const findManySpy = jest.spyOn(prisma.user, "findMany").mockResolvedValue(mockUsers);

      const result = await userService.getUsers();

      expect(result).toEqual(mockUsersWithNoPassword);
      expect(findManySpy).toHaveBeenCalledTimes(1);
      expect(result).toMatchSnapshot();
    });

    test("getUser は指定されたIDのユーザーを返す", async () => {
      const userId = 1;
      const findUniqueSpy = jest.spyOn(prisma.user, "findUnique").mockResolvedValue(mockUser);

      const result = await userService.getUser(userId);

      expect(result).toEqual({ id: mockUser.id, name: mockUser.name, email: mockUser.email });
      expect(findUniqueSpy).toHaveBeenCalledWith({
        where: { id: userId },
      });
      expect(result).toMatchSnapshot();
    });

    test("updateUser は指定されたIDのユーザーを更新して返す", async () => {
      const userId = 1;
      const newName = "Alice Updated";
      const updatedUser = {
        id: userId,
        name: newName,
        email: "test@test.com",
        passwordHash: "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
      };
      const updateSpy = jest.spyOn(prisma.user, "update").mockResolvedValue(updatedUser);

      const result = await userService.updateUser(userId, newName);

      const { passwordHash, ...publicUser } = updatedUser;

      expect(result).toEqual(publicUser);
      expect(updateSpy).toHaveBeenCalledWith({
        where: { id: userId },
        data: { name: newName },
      });
      expect(result).toMatchSnapshot();
    });

    test("deleteUser は指定されたIDのユーザーを削除する", async () => {
      const userId = 1;
      const deleteSpy = jest.spyOn(prisma.user, "delete").mockResolvedValue(mockUser);

      await userService.deleteUser(userId);

      expect(deleteSpy).toHaveBeenCalledWith({
        where: { id: userId },
      });
    });
  });

  describe("異常系", () => {
    test("getUserで存在しないIDを指定した場合、nullを返す", async () => {
      const userId = 999;
      const findUniqueSpy = jest.spyOn(prisma.user, "findUnique").mockResolvedValue(null);

      const result = await userService.getUser(userId);

      expect(result).toBeNull();
      expect(findUniqueSpy).toHaveBeenCalledWith({
        where: { id: userId },
      });
    });

    test("updateUserで存在しないIDを指定した場合、エラーになる", async () => {
      const userId = 999;
      const newName = "Alice Updated";
      const updateSpy = jest.spyOn(prisma.user, "update").mockRejectedValue(new AppError(404, "User not found"));

      await expect(userService.updateUser(userId, newName)).rejects.toMatchObject({
        statusCode: 404,
        message: "User not found",
      });
      expect(updateSpy).toHaveBeenCalledTimes(1);
      expect(updateSpy).toHaveBeenCalledWith({
        where: { id: userId },
        data: { name: newName },
      });
    });

    test("deleteUserで存在しないIDを指定した場合、エラーになる", async () => {
      const userId = 999;
      const deleteSpy = jest.spyOn(prisma.user, "delete").mockRejectedValue(new AppError(404, "User not found"));
      await expect(userService.deleteUser(userId)).rejects.toMatchObject({
        statusCode: 404,
        message: "User not found",
      });
      expect(deleteSpy).toHaveBeenCalledTimes(1);
      expect(deleteSpy).toHaveBeenCalledWith({
        where: { id: userId },
      });
    });
  });
});
