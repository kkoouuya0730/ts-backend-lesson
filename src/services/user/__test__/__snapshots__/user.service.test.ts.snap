// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`User Service 正常系 getUser は指定されたIDのユーザーを返す 1`] = `
{
  "email": "test1@test.com",
  "id": 1,
  "name": "Alice",
  "role": "user",
}
`;

exports[`User Service 正常系 getUsers はユーザー一覧を返す 1`] = `
[
  {
    "email": "test1@test.com",
    "id": 1,
    "name": "Alice",
    "role": "user",
  },
  {
    "email": "test2@test.com",
    "id": 2,
    "name": "Bob",
    "role": "user",
  },
]
`;

exports[`User Service 正常系 updateUser は指定されたIDのユーザーを更新して返す 1`] = `
{
  "email": "test@test.com",
  "id": 1,
  "name": "Alice Updated",
  "role": "user",
}
`;
